src/dbnode/persist/fs/persist_manager.go

```diff
@@ -644,3 +670,28 @@ func (pm *persistManager) SetRuntimeOptions(value runtime.Options) {
 	pm.currRateLimitOpts = value.PersistRateLimitOptions()
 	pm.Unlock()
 }
+
+// Ensure only a single index writer is ever created for a block, returns an error if an index writer
+// already exists.
+func (pm *persistManager) ensureSingleIndexWriterForBlock(blockStart time.Time) (*singleUseIndexWriter, error) {
+	pm.Lock()
+	defer pm.Unlock()
+
+	idxWriter, ok := pm.indexPM.writersByBlockStart[xtime.ToUnixNano(blockStart)]
+	// Ensure that we have not already created an index writer for this block.
+	if ok {
+		return nil, errPersistManagerIndexWriterAlreadyExists
+	}
+
+	writer, err := pm.indexPM.newIndexWriterFn(pm.opts)
+	if err != nil {
+		return nil, err
+	}
+	idxWriter = &singleUseIndexWriter{
+		manager: &pm.indexPM,
+		writer:  writer,
+	}
+	pm.indexPM.writersByBlockStart[xtime.ToUnixNano(blockStart)] = idxWriter
```

Comment:
Instead of caching these per block (which would require us to age out and invalidate results from cache) should we just allocate each time?

It doesn't look like these are that big, they just have a pointer to the writer and manager, the only thing they really create is the index writer which is fairly small.